Riot/Modules/Room/VoiceMessages/VoiceMessageAttachmentCacheManager.swift

```diff
@@ -0,0 +1,201 @@
+// 
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+import DSWaveformImage
+
+enum VoiceMessageAttachmentCacheManagerError: Error {
+    case invalidEventId
+    case invalidAttachmentType
+    case decryptionError(Error)
+    case preparationError(Error)
+    case conversionError(Error)
+    case invalidNumberOfSamples
+    case samplingError
+}
+
+/**
+ Swift optimizes the callbacks to be the same instance. Wrap them so we can store them in an array.
+ */
+private class CompletionWrapper {
+    let completion: (Result<(String, URL, TimeInterval, [Float]), Error>) -> Void
+    
+    init(_ completion: @escaping (Result<(String, URL, TimeInterval, [Float]), Error>) -> Void) {
+        self.completion = completion
+    }
+}
+
+class VoiceMessageAttachmentCacheManager {
+    
+    static let sharedManager = VoiceMessageAttachmentCacheManager()
+    
+//    private let workQueue: DispatchQueue
+    
+    private var completionCallbacks = [String: [CompletionWrapper]]()
+    private var samples = [String: [Int: [Float]]]()
+    private var durations = [String: TimeInterval]()
+    private var finalURLs = [String: URL]()
+    
+//    private init() {
```

Comment:
It may be a good idea to uncomment `private init` method because the class is intended to be used as a singleton.